src/t_hash.c

```diff
@@ -2266,29 +2963,984 @@ void hpersistCommand(client *c) {
+    if (*firstFieldPos < 0) {
+        addReplyError(c, "missing FIELDS argument");
+        return C_ERR;
+    }
+
+    if (*flags & HFE_CMD_COND_MASK &&
+        (!(*flags & (HFE_CMD_EX | HFE_CMD_EXAT | HFE_CMD_PX | HFE_CMD_PXAT))))
+    {
+        addReplyError(c, "NX, XX, GT, and LT can be specified only when EX, PX, EXAT, or PXAT is specified");
+        return C_ERR;
+    }
+
+    return C_OK;
+
+err_missing_expire:
+    addReplyError(c, "missing expire time");
+    return C_ERR;
+err_condition:
+    addReplyError(c, "Only one of NX, XX, GT, and LT arguments can be specified");
+    return C_ERR;
+err_expiration:
+    addReplyError(c, "Only one of EX, PX, EXAT, PXAT or PERSIST arguments can be specified");
+    return C_ERR;
+}
+
+/* Reply with field value and optionally set expire time according to 'flag'.
+ * Return 1 if expire time is updated. */
+static int hgetfReplyValueAndSetExpiry(client *c, robj *o, sds field, int flag,
+                                       uint64_t expireAt, uint64_t *minPrevExp)
+{
+    unsigned char *fptr = NULL, *vptr = NULL, *tptr, *h;
+    hfield hf = NULL;
+    dict *d = NULL;
+    dictEntry *de = NULL;
+    uint64_t prevExpire = EB_EXPIRE_TIME_INVALID;
+
+    if (o->encoding == OBJ_ENCODING_HT) {
+        d = o->ptr;
+        /* First retrieve the field to check if it exists */
+        de = dictFind(d, field);
         if (de == NULL) {
-            addReplyLongLong(c, HFE_PERSIST_NO_FIELD);
-            continue;
+            addReplyNull(c);
+            return 0;
         }
 
-        hfield hf = dictGetKey(de);
-        uint64_t expire = hfieldGetExpireTime(hf);
-        if (expire == EB_EXPIRE_TIME_INVALID) {
-            addReplyLongLong(c, HFE_PERSIST_NO_TTL);
-            continue;
+        hf = dictGetKey(de);
+        if (hfieldIsExpired(hf)) {
+            addReplyNull(c);
+            return 0;
         }
+        prevExpire = hfieldGetExpireTime(hf);
 
-        /* Already expired. Pretend there is no such field */
-        if ( (long long) expire < commandTimeSnapshot()) {
-            addReplyLongLong(c, HFE_PERSIST_NO_FIELD);
-            continue;
+        /* Reply with value */
+        sds val = dictGetVal(de);
+        addReplyBulkCBuffer(c, val, sdslen(val));
+    } else if (o->encoding == OBJ_ENCODING_LISTPACK_EX) {
+        long long expire;
+        unsigned char *vstr = NULL;
+        unsigned int vlen = UINT_MAX;
+        long long vll = LLONG_MAX;
+        listpackEx *lpt = o->ptr;
+
+        fptr = lpFirst(lpt->lp);
+        if (fptr != NULL) {
+            fptr = lpFind(lpt->lp, fptr, (unsigned char *) field, sdslen(field), 2);
+            if (fptr != NULL) {
+                vptr = lpNext(lpt->lp, fptr);
+                serverAssert(vptr != NULL);
+
+                tptr = lpNext(lpt->lp, vptr);
+                serverAssert(tptr != NULL);
+
+                h = lpGetValue(tptr, NULL, &expire);
+                serverAssert(h == NULL);
```

Comment:
ref https://github.com/redis/redis/pull/13243#discussion_r1602571987

this assertiong will never be triggered.
you passed NULL to the second parameter of `lpGetValue()`, if the field is string, it will crash in `*slen = ele_len;`.
```c
    vstr = lpGet(p, &ele_len, NULL);
    if (vstr) {
        *slen = ele_len;          <- here
    } else {
        *lval = ele_len;
    }
```

crash log:
```
=== REDIS BUG REPORT START: Cut & paste starting from here ===
64716:M 16 May 2024 11:55:33.528 # Redis 255.255.255 crashed by signal: 11, si_code: 1
64716:M 16 May 2024 11:55:33.528 # Accessing address: (nil)
64716:M 16 May 2024 11:55:33.528 # Crashed running the instruction at: 0x55c8922a2fa9

------ STACK TRACE ------
EIP:
./src/redis-server *:6379(lpGetValue+0x54)[0x55c8922a2fa9]

64716 redis-server *
/lib/x86_64-linux-gnu/libc.so.6(+0x45320)[0x701fc8a45320]
./src/redis-server *:6379(lpGetValue+0x54)[0x55c8922a2fa9]
./src/redis-server *:6379(+0x11140f)[0x55c89220b40f]
./src/redis-server *:6379(hgetfCommand+0x1a3)[0x55c89220b987]
./src/redis-server *:6379(call+0x143)[0x55c89219cb67]
./src/redis-server *:6379(processCommand+0xea1)[0x55c89219e6e1]
./src/redis-server *:6379(processCommandAndResetClient+0x39)[0x55c8921bbcc9]
./src/redis-server *:6379(processInputBuffer+0x1a0)[0x55c8921bbf4e]
./src/redis-server *:6379(readQueryFromClient+0x4b5)[0x55c8921bc516]
./src/redis-server *:6379(+0x1b836f)[0x55c8922b236f]
./src/redis-server *:6379(+0x1b8a98)[0x55c8922b2a98]
./src/redis-server *:6379(aeProcessEvents+0x24d)[0x55c892186438]
./src/redis-server *:6379(aeMain+0x2e)[0x55c892186676]
./src/redis-server *:6379(main+0xd5a)[0x55c8921a7ebe]
/lib/x86_64-linux-gnu/libc.so.6(+0x2a1ca)[0x701fc8a2a1ca]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x8b)[0x701fc8a2a28b]
./src/redis-server *:6379(_start+0x25)[0x55c892180155]
```
what about adding a new method like `int lpGetInteger(unsigned char *lp, long long *lval)`, return 1 if it's a integer.

Reply:
how did you trigger this crash?

Reply:
https://github.com/sundb/redis/commit/17b544f3530b0ce81c7e4180b6e0c4c9df29e85f
it's easy to reproduce the corrupt by using the patch.
then run `./runtest --single integration/corrupt-dump-fuzzer --accurate`
@tezc